README.md
# ml-kmeans

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![npm download][download-image]][download-url]

[K-means clustering][] aims to partition n observations into k clusters in which
each observation belongs to the cluster with the nearest mean.

## Installation

`npm i ml-kmeans`

## [API Documentation](https://mljs.github.io/kmeans/)

## Example

```js
const kmeans = require('ml-kmeans');

let data = [[1, 1, 1], [1, 2, 1], [-1, -1, -1], [-1, -1, -1.5]];
let centers = [[1, 2, 1], [-1, -1, -1]];

let ans = kmeans(data, 2, { initialization: centers });
console.log(ans);
/*
KMeansResult {
  clusters: [ 0, 0, 1, 1 ],
  centroids: 
   [ { centroid: [ 1, 1.5, 1 ], error: 0.25, size: 2 },
     { centroid: [ -1, -1, -1.25 ], error: 0.0625, size: 2 } ],
  converged: true,
  iterations: 1
}
*/
```

## Authors

- [Miguel Asencio](https://github.com/maasencioh)

## Sources

D. Arthur, S. Vassilvitskii, k-means++: The Advantages of Careful Seeding, in: Proc. of the 18th Annual
ACM-SIAM Symposium on Discrete Algorithms, 2007, pp. 1027–1035.
[Link to article](http://ilpubs.stanford.edu:8090/778/1/2006-13.pdf)

## License

[MIT](./LICENSE)

[npm-image]: https://img.shields.io/npm/v/ml-kmeans.svg?style=flat-square
[npm-url]: https://npmjs.org/package/ml-kmeans
[travis-image]: https://img.shields.io/travis/mljs/kmeans/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/mljs/kmeans
[codecov-image]: https://img.shields.io/codecov/c/github/mljs/kmeans.svg?style=flat-square
[codecov-url]: https://codecov.io/github/mljs/kmeans
[download-image]: https://img.shields.io/npm/dm/ml-kmeans.svg?style=flat-square
[download-url]: https://npmjs.org/package/ml-kmeans
[k-means clustering]: https://en.wikipedia.org/wiki/K-means_clustering


History.md
# [5.0.0](https://github.com/mljs/kmeans/compare/v4.2.1...v5.0.0) (2019-06-29)


### chore

* update dependencies and remove support for Node.js 6 ([5efe565](https://github.com/mljs/kmeans/commit/5efe565))


### BREAKING CHANGES

* Node.js 6 is no longer supported.



<a name="4.2.1"></a>
## [4.2.1](https://github.com/mljs/kmeans/compare/v4.2.0...v4.2.1) (2018-08-15)



<a name="4.2.0"></a>
# [4.2.0](https://github.com/mljs/kmeans/compare/v4.1.0...v4.2.0) (2018-08-15)



<a name="4.1.0"></a>
# [4.1.0](https://github.com/mljs/kmeans/compare/v4.0.1...v4.1.0) (2018-08-11)



<a name="4.0.1"></a>
## [4.0.1](https://github.com/mljs/kmeans/compare/v4.0.0...v4.0.1) (2018-05-24)


### Bug Fixes

* keep initial value for empty centroids ([94b0de9](https://github.com/mljs/kmeans/commit/94b0de9))



<a name="4.0.0"></a>
# [4.0.0](https://github.com/mljs/kmeans/compare/v3.1.0...v4.0.0) (2018-05-23)


### Features

* implement kmeans++ initialization algorithm ([007f3b1](https://github.com/mljs/kmeans/commit/007f3b1))


### BREAKING CHANGES

* kmeans++ is now the default initialization algorithm
instead of mostDistant



<a name="3.1.0"></a>
# [3.1.0](https://github.com/mljs/kmeans/compare/v3.0.1...v3.1.0) (2018-05-16)



<a name="3.0.1"></a>
## [3.0.1](https://github.com/mljs/kmeans/compare/v3.0.0...v3.0.1) (2018-05-04)



<a name="3.0.0"></a>
# [3.0.0](https://github.com/mljs/kmeans/compare/v2.0.0...v3.0.0) (2016-10-07)


### Bug Fixes

* remove unneeded array creations in updateClusterID ([b2957d5](https://github.com/mljs/kmeans/commit/b2957d5))


### Features

* accepts distance function as parameter ([68993a1](https://github.com/mljs/kmeans/commit/68993a1))
* allows to compute a new array of points their cluster id ([254f9a9](https://github.com/mljs/kmeans/commit/254f9a9))
* return a generator in the withIterations option ([3b4c7c6](https://github.com/mljs/kmeans/commit/3b4c7c6))
* return the number of iteration done ([75fd42e](https://github.com/mljs/kmeans/commit/75fd42e))
* returns the error and the size of each cluster ([e85aab7](https://github.com/mljs/kmeans/commit/e85aab7))
* when maxIterations is 0, iterates until converge ([220c828](https://github.com/mljs/kmeans/commit/220c828))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/mljs/kmeans/compare/v1.0.0...v2.0.0) (2016-09-27)


### Features

* initialize the points as parameter ([fbe993d](https://github.com/mljs/kmeans/commit/fbe993d))
* **initialization:** moreDistant initialization ([d098f43](https://github.com/mljs/kmeans/commit/d098f43))
* **initialization:** random initialization ([9e58baf](https://github.com/mljs/kmeans/commit/9e58baf))


### BREAKING CHANGES

* instead of asking for a manual initialization, ask for k and the initialization method



<a name="1.0.0"></a>
# [1.0.0](https://github.com/mljs/kmeans/compare/v0.1.0...v1.0.0) (2016-09-15)


### Chores

* update package ([10dd6be](https://github.com/mljs/kmeans/commit/10dd6be))


### BREAKING CHANGES

* rename option variables to more understandable names



<a name="0.1.0"></a>
# 0.1.0 (2015-07-22)





src/kmeans.js
import { squaredEuclidean } from 'ml-distance-euclidean';

import {
  updateClusterID,
  updateCenters,
  hasConverged,
  calculateDistanceMatrix
} from './utils';
import { mostDistant, random, kmeanspp } from './initialization';
import KMeansResult from './KMeansResult';

const defaultOptions = {
  maxIterations: 100,
  tolerance: 1e-6,
  withIterations: false,
  initialization: 'kmeans++',
  distanceFunction: squaredEuclidean
};

/**
 * Each step operation for kmeans
 * @ignore
 * @param {Array<Array<number>>} centers - K centers in format [x,y,z,...]
 * @param {Array<Array<number>>} data - Points [x,y,z,...] to cluster
 * @param {Array<number>} clusterID - Cluster identifier for each data dot
 * @param {number} K - Number of clusters
 * @param {object} [options] - Option object
 * @param {number} iterations - Current number of iterations
 * @return {KMeansResult}
 */
function step(centers, data, clusterID, K, options, iterations) {
  clusterID = updateClusterID(
    data,
    centers,
    clusterID,
    options.distanceFunction
  );
  var newCenters = updateCenters(centers, data, clusterID, K);
  var converged = hasConverged(
    newCenters,
    centers,
    options.distanceFunction,
    options.tolerance
  );
  return new KMeansResult(
    clusterID,
    newCenters,
    converged,
    iterations,
    options.distanceFunction
  );
}

/**
 * Generator version for the algorithm
 * @ignore
 * @param {Array<Array<number>>} centers - K centers in format [x,y,z,...]
 * @param {Array<Array<number>>} data - Points [x,y,z,...] to cluster
 * @param {Array<number>} clusterID - Cluster identifier for each data dot
 * @param {number} K - Number of clusters
 * @param {object} [options] - Option object
 */
function* kmeansGenerator(centers, data, clusterID, K, options) {
  var converged = false;
  var stepNumber = 0;
  var stepResult;
  while (!converged && stepNumber < options.maxIterations) {
    stepResult = step(centers, data, clusterID, K, options, ++stepNumber);
    yield stepResult.computeInformation(data);
    converged = stepResult.converged;
    centers = stepResult.centroids;
  }
}

/**
 * K-means algorithm
 * @param {Array<Array<number>>} data - Points in the format to cluster [x,y,z,...]
 * @param {number} K - Number of clusters
 * @param {object} [options] - Option object
 * @param {number} [options.maxIterations = 100] - Maximum of iterations allowed
 * @param {number} [options.tolerance = 1e-6] - Error tolerance
 * @param {boolean} [options.withIterations = false] - Store clusters and centroids for each iteration
 * @param {function} [options.distanceFunction = squaredDistance] - Distance function to use between the points
 * @param {number} [options.seed] - Seed for random initialization.
 * @param {string|Array<Array<number>>} [options.initialization = 'kmeans++'] - K centers in format [x,y,z,...] or a method for initialize the data:
 *  * You can either specify your custom start centroids, or select one of the following initialization method:
 *  * `'kmeans++'` will use the kmeans++ method as described by http://ilpubs.stanford.edu:8090/778/1/2006-13.pdf
 *  * `'random'` will choose K random different values.
 *  * `'mostDistant'` will choose the more distant points to a first random pick
 * @return {KMeansResult} - Cluster identifier for each data dot and centroids with the following fields:
 *  * `'clusters'`: Array of indexes for the clusters.
 *  * `'centroids'`: Array with the resulting centroids.
 *  * `'iterations'`: Number of iterations that took to converge
 */
export default function kmeans(data, K, options) {
  options = Object.assign({}, defaultOptions, options);

  if (K <= 0 || K > data.length || !Number.isInteger(K)) {
    throw new Error(
      'K should be a positive integer smaller than the number of points'
    );
  }

  var centers;
  if (Array.isArray(options.initialization)) {
    if (options.initialization.length !== K) {
      throw new Error('The initial centers should have the same length as K');
    } else {
      centers = options.initialization;
    }
  } else {
    switch (options.initialization) {
      case 'kmeans++':
        centers = kmeanspp(data, K, options);
        break;
      case 'random':
        centers = random(data, K, options.seed);
        break;
      case 'mostDistant':
        centers = mostDistant(
          data,
          K,
          calculateDistanceMatrix(data, options.distanceFunction),
          options.seed
        );
        break;
      default:
        throw new Error(
          `Unknown initialization method: "${options.initialization}"`
        );
    }
  }

  // infinite loop until convergence
  if (options.maxIterations === 0) {
    options.maxIterations = Number.MAX_VALUE;
  }

  var clusterID = new Array(data.length);
  if (options.withIterations) {
    return kmeansGenerator(centers, data, clusterID, K, options);
  } else {
    var converged = false;
    var stepNumber = 0;
    var stepResult;
    while (!converged && stepNumber < options.maxIterations) {
      stepResult = step(centers, data, clusterID, K, options, ++stepNumber);
      converged = stepResult.converged;
      centers = stepResult.centroids;
    }
    return stepResult.computeInformation(data);
  }
}
